### PR TITLE
Day 19 - Service, Broadcast Receiver, Notification

### PR DESCRIPTION
**Pull request by haingo94**
#### Related issues (Optional)
Task #100

#### Description (Required)
- [x] Play a list of songs (online )
- [x] Have function: `Next` - `Previous` - `Play` - `Shuffle` - `Auto Next`
- [x] When press `Lockscreen` -> Show notification with:
         - Image's song, name's song, name's singer
         - Buton `close notification`, and when press it -> stop `play music`
         - Song's progress ( show song's duration and current time )

### CODE REVIEW

File: app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            }
+        });
+    }
+
+    @Override
+    public int getItemCount() {
+        return mSongs.size();
+    }
+
+    class ItemViewHolder extends RecyclerView.ViewHolder {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        ItemViewHolder(View itemView) {
+            super(itemView);
+            mTvTitle = (TextView) itemView.findViewById(R.id.tvSongTitle);
+            mTvArtist = (TextView) itemView.findViewById(R.id.tvArtist);
+            mTvTime = (TextView) itemView.findViewById(R.id.tvTime);
+        }
+    }
+
+    public interface OnListener{
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java]: ERROR: Missing a Javadoc comment.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+                        mCurrentPosition++;
+                    }
+                }
+                playSong(mSongs, mCurrentPosition);
+                sendBroadcastSong(Action.NEXT_SONG.getValue());
+
+            } else if (intent.getAction().equals(Action.PAUSE.getValue())) {
+                if (mMediaPLayer.isPlaying()) {
+                    mMediaPLayer.pause();
+                    Log.d("xxx", "onStartCommand: pause " + mMediaPLayer.isPlaying());
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [AvoidDuplicateLiterals] The String literal "xxx" appears 5 times in this file; the first occurrence is on line 84
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                .addAction(R.drawable.ic_skip_next_black_24dp, null, nextPendingIntent)
+                .build();
+        startForeground(1, builder);
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        mMediaPLayer.stop();
+        stopForeground(true);
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            playIntent.setAction(Action.PAUSE.getValue());
+        } else {
+            Log.d("xxx", "showNotification: nononono");
+            playIntent.setAction(Action.RESUME.getValue());
+        }
+        PendingIntent playPendingIntent = PendingIntent.getService(this, 0, playIntent, 0);
+
+        Notification builder = new NotificationCompat.Builder(this)
+                .setContent(remoteViews)
+                .setSmallIcon(R.drawable.ic_play_arrow_black_24dp)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [IconColors] Notification icons must be entirely white
 Notification icons and Action Bar icons should only white and shades of gray. See the Android Design Guide for more details. Note that the way Lint decides whether an icon is an action bar icon or a notification icon is based on the filename prefix: `ic_menu_` for action bar icons, `ic_stat_` for notification icons etc. These correspond to the naming conventions documented in http://developer.android.com/guide/practices/ui_guidelines/icon_design.html

File: app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        ItemViewHolder(View itemView) {
+            super(itemView);
+            mTvTitle = (TextView) itemView.findViewById(R.id.tvSongTitle);
+            mTvArtist = (TextView) itemView.findViewById(R.id.tvArtist);
+            mTvTime = (TextView) itemView.findViewById(R.id.tvTime);
+            mImgIcon = (ImageView) itemView.findViewById(R.id.imgIcon);
+        }
+    }
+
+    public interface OnListener{
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java]: ERROR: Missing a Javadoc comment.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+                        mCurrentPosition++;
+                    }
+                }
+                playSong(mSongs, mCurrentPosition);
+                sendBroadcastSong(Action.NEXT_SONG.getValue());
+
+            } else if (intent.getAction().equals(Action.PAUSE.getValue())) {
+                if (mMediaPLayer.isPlaying()) {
+                    mMediaPLayer.pause();
+                    Log.d("xxx", "onStartCommand: pause " + mMediaPLayer.isPlaying());
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [AvoidDuplicateLiterals] The String literal "xxx" appears 6 times in this file; the first occurrence is on line 84
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

File: app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            }
+        });
+    }
+
+    @Override
+    public int getItemCount() {
+        return mSongs.size();
+    }
+
+    class ItemViewHolder extends RecyclerView.ViewHolder {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java]: ERROR: Missing a Javadoc comment.

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java (hunk 7 above)
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [AvoidDuplicateLiterals] The String literal "xxx" appears 6 times in this file; the first occurrence is on line 84
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+                .addAction(R.drawable.ic_skip_next_black_24dp, null, nextPendingIntent)
+                .build();
+        startForeground(1, builder);
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        mMediaPLayer.stop();
+        stopForeground(true);
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+            playIntent.setAction(Action.PAUSE.getValue());
+        } else {
+            Log.d("xxx", "showNotification: nononono");
+            playIntent.setAction(Action.RESUME.getValue());
+        }
+        PendingIntent playPendingIntent = PendingIntent.getService(this, 0, playIntent, 0);
+
+        Notification builder = new NotificationCompat.Builder(this)
+                .setContent(remoteViews)
+                .setSmallIcon(R.drawable.ic_play_arrow_black_24dp)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [IconColors] Notification icons must be entirely white
 Notification icons and Action Bar icons should only white and shades of gray. See the Android Design Guide for more details. Note that the way Lint decides whether an icon is an action bar icon or a notification icon is based on the filename prefix: `ic_menu_` for action bar icons, `ic_stat_` for notification icons etc. These correspond to the naming conventions documented in http://developer.android.com/guide/practices/ui_guidelines/icon_design.html

File: app/src/main/java/vn/asiantech/internship/models/Song.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+ */
+
+public class Song implements Parcelable {
+    private int id;
+    private String title;
+    private String artist;
+    private String url;
+    private String imageUrl;
+    private int time;
+    private boolean isPlaying;
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/models/Song.java]: WARNING: [AvoidFieldNameMatchingMethodName] Field isPlaying has the same name as a method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/naming.html#AvoidFieldNameMatchingMethodName

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mCountDownTimer != null) {
+            mCountDownTimer.cancel();
+        }
+        mMediaPLayer.stop();
+        stopForeground(true);
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/ui/music/SongPlayingActivity.java [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+        Intent intent = new Intent(SongPlayingActivity.this, MusicService.class);
+        intent.putExtra(Action.SEEK_TO.getValue(), seekBar.getProgress());
+        intent.setAction(Action.SEEK_TO.getValue());
+        sendBroadcast(intent);
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        unregisterReceiver(mBroadcastReceiver);
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/ui/music/SongPlayingActivity.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/res/anim/rotate_disk.xml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
```
Comment (at-dev-support-01) [inline app/src/main/res/anim/rotate_disk.xml]: WARNING: [UnusedResources] The resource `R.anim.rotate_disk` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+
+        Intent previousIntent = new Intent(Action.PREVIOUS_SONG.getValue());
+        PendingIntent previousPendingIntent = PendingIntent.getBroadcast(this, 0, previousIntent, 0);
+
+        Intent closeNotificationIntent = new Intent(Action.CLOSE.getValue());
+        PendingIntent closeNotificationPendingIntent = PendingIntent.getBroadcast(this, 0, closeNotificationIntent, 0);
+
+        Notification builder = new NotificationCompat.Builder(this)
+                .setContent(remoteViews)
+                .setSmallIcon(R.drawable.ic_play_arrow_black_24dp)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [IconColors] Notification icons must be entirely white
 Notification icons and Action Bar icons should only white and shades of gray. See the Android Design Guide for more details. Note that the way Lint decides whether an icon is an action bar icon or a notification icon is based on the filename prefix: `ic_menu_` for action bar icons, `ic_stat_` for notification icons etc. These correspond to the naming conventions documented in http://developer.android.com/guide/practices/ui_guidelines/icon_design.html

File: app/src/main/res/anim/rotate_disk.xml [hunk 16]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<rotate xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (at-dev-support-01) [inline app/src/main/res/anim/rotate_disk.xml]: WARNING: [UnusedResources] The resource `R.anim.rotate_disk` appears to be unused
 Unused resources make applications larger and slow down builds.

**Comment by at-vinhhuynh** [comment]
@haingo94 
Please fix CI comment first

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mCountDownTimer != null) {
+            mCountDownTimer.cancel();
+        }
+        mMediaPLayer.stop();
+//        unregisterReceiver(mBroadcastReceiver);
+        stopForeground(true);
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ *
+ * Created by quanghai on 30/06/2017.
+ */
+public class SongAdapter extends RecyclerView.Adapter<SongAdapter.ItemViewHolder> {
+    private List<Song> mSongs = new ArrayList<>();
+    private OnSelectSongListener mOnSelectSongListener;
+    private SongPlayingActivity mActivity = new SongPlayingActivity();
+
+    public SongAdapter(List<Song> mSongs, OnSelectSongListener onSelectSongListener) {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java]: `mSongs` to `songs`

File: app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+
+    /**
+     * Created by quanghai on 30/06/2017.
+     */
+    public interface OnSelectSongListener {
+        void onItemClick(int position);
+    }
+}
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/adapters/SongAdapter.java]: add a break line

File: app/src/main/java/vn/asiantech/internship/models/Song.java [hunk 20]
```diff
@@ -0,0 +1,134 @@
+package vn.asiantech.internship.models;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+/**
+ * Created by quanghai on 30/06/2017.
+ */
+
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/models/Song.java]: remove thís line

File: app/src/main/java/vn/asiantech/internship/services/MusicService.java [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mCountDownTimer != null) {
+            mCountDownTimer.cancel();
+        }
+        mMediaPLayer.stop();
+//        unregisterReceiver(mBroadcastReceiver);
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/services/MusicService.java]: why comment this line?

File: app/src/main/res/layout/activity_song_detail.xml [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+        android:layout_below="@+id/tvArtist"
+        android:layout_centerHorizontal="true"
+        android:layout_marginTop="@dimen/imageview.disk.margin.top"
+        android:src="@drawable/img_disk" />
+
+    <SeekBar
+        android:id="@+id/seekBarTime"
+        android:layout_width="match_parent"
+        android:layout_height="5dp"
+
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/activity_song_detail.xml]: remove this line

File: app/src/main/res/layout/activity_song_detail.xml [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+        android:layout_below="@+id/imgDisk"
+        android:layout_marginTop="@dimen/seekbat.margin.top" />
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_below="@+id/seekBarTime">
+
+        <TextView
+            android:id="@+id/tvcurrentTime"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/activity_song_detail.xml]: `tvcurrentTime` to `tvCurrentTime`

**Comment by haingo94** [comment]
@at-vinhhuynh thank you so much

**Pull request merged**